pkg/frontend/querymiddleware/error_caching_test.go

```diff
@@ -22,200 +22,263 @@ import (
+				require.Equal(t, 0, c.CountStoreCalls())
+			})
+
+			t.Run("disabled by option", func(t *testing.T) {
+				c := cache.NewInstrumentedMockCache()
+
+				innerRes := newEmptyPrometheusResponse()
+				inner := &mockHandler{}
+				inner.On("Do", mock.Anything, mock.Anything).Return(innerRes, nil)
+
+				ctx := user.InjectOrgID(context.Background(), "1234")
+				req := tc.newQueryRequest(Options{
+					CacheDisabled: true,
+				})
+				res, err := runHandler(ctx, inner, c, req)
+
+				require.NoError(t, err)
+				require.Equal(t, innerRes, res)
+				require.Equal(t, 0, c.CountFetchCalls())
+				require.Equal(t, 0, c.CountStoreCalls())
+			})
+
+			t.Run("cache hit", func(t *testing.T) {
+				c := cache.NewInstrumentedMockCache()
+
+				inner := &mockHandler{}
+
+				ctx := user.InjectOrgID(context.Background(), "1234")
+				req := tc.newQueryRequest(Options{})
+				key := keyGen.QueryRequestError(ctx, "1234", req)
+				bytes, err := proto.Marshal(&CachedError{
+					Key:          key,
+					ErrorType:    string(apierror.TypeExec),
+					ErrorMessage: "limits error",
+				})
+				require.NoError(t, err)
+
+				// NOTE: We rely on this mock cache being synchronous
+				c.SetAsync(cacheHashKey(key), bytes, time.Minute)
+				res, err := runHandler(ctx, inner, c, req)
+
+				require.Error(t, err)
+				require.Nil(t, res)
+				require.Equal(t, 1, c.CountFetchCalls())
+				require.Equal(t, 1, c.CountStoreCalls())
+			})
+
+			t.Run("cache hit key collision", func(t *testing.T) {
+				c := cache.NewInstrumentedMockCache()
+
+				innerRes := newEmptyPrometheusResponse()
+				inner := &mockHandler{}
+				inner.On("Do", mock.Anything, mock.Anything).Return(innerRes, nil)
+
+				ctx := user.InjectOrgID(context.Background(), "1234")
+				req := tc.newQueryRequest(Options{})
+
+				key := keyGen.QueryRequestError(ctx, "1234", req)
+				bytes, err := proto.Marshal(&CachedError{
+					Key:          "different key that is stored under the same hashed key",
+					ErrorType:    string(apierror.TypeExec),
+					ErrorMessage: "limits error",
+				})
+				require.NoError(t, err)
+
+				// NOTE: We rely on this mock cache being synchronous
+				c.SetAsync(cacheHashKey(key), bytes, time.Minute)
+				res, err := runHandler(ctx, inner, c, req)
+
+				require.NoError(t, err)
+				require.Equal(t, innerRes, res)
+				require.Equal(t, 1, c.CountFetchCalls())
+				require.Equal(t, 1, c.CountStoreCalls())
+			})
+
+			t.Run("corrupt cache data", func(t *testing.T) {
+				c := cache.NewInstrumentedMockCache()
+
+				innerRes := newEmptyPrometheusResponse()
+				inner := &mockHandler{}
+				inner.On("Do", mock.Anything, mock.Anything).Return(innerRes, nil)
+
+				ctx := user.InjectOrgID(context.Background(), "1234")
+				req := tc.newQueryRequest(Options{})
+
+				key := keyGen.QueryRequestError(ctx, "1234", req)
+				bytes := []byte{0x0, 0x0, 0x0, 0x0}
+
+				// NOTE: We rely on this mock cache being synchronous
+				c.SetAsync(cacheHashKey(key), bytes, time.Minute)
```

Comment:
The `Cache` interface has grown a synchronous `.Set()` method since this test was written.